x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/BucketCorrelationAggregationIT.java

```diff
@@ -0,0 +1,228 @@
+        Terms.Bucket dogBucket = terms.getBucketByKey("dog");
+        NumericMetricsAggregation.SingleValue approxCatCorrelation = catBucket.getAggregations().get("correlates");
+        NumericMetricsAggregation.SingleValue approxDogCorrelation = dogBucket.getAggregations().get("correlates");
+
+        assertThat(approxCatCorrelation.value(), closeTo(catCorrelation, 0.1));
+        assertThat(approxDogCorrelation.value(), closeTo(dogCorrelation, 0.1));
+    }
+
+    private static Tuple<RangeAggregationBuilder, BucketCorrelationAggregationBuilder> buildRangeAggAndSetExpectations(
+        Percentiles raw_percentiles,
+        double[] steps,
+        long totalCount,
+        String indicatorFieldName
+    ) {
+        List<Double> percentiles = new ArrayList<>();
+        List<Double> fractions = new ArrayList<>();
+        RangeAggregationBuilder builder = AggregationBuilders.range("correlation_range").field(indicatorFieldName);
+        double percentile_0 = raw_percentiles.percentile(steps[0]);
+        builder.addUnboundedTo(percentile_0);
+        fractions.add(0.02);
+        percentiles.add(percentile_0);
+        int last_added = 0;
+        for (int i = 1; i < steps.length; i++) {
+            double percentile_l = raw_percentiles.percentile(steps[i - 1]);
+            double percentile_r = raw_percentiles.percentile(steps[i]);
+            if (Double.compare(percentile_l, percentile_r) == 0) {
+                fractions.set(last_added, fractions.get(last_added) + 0.02);
+            } else {
+                last_added = i;
+                fractions.add(0.02);
+                percentiles.add(percentile_r);
+            }
+        }
+        fractions.add(2.0/100);
+        double[] expectations = new double[percentiles.size() + 1];
+        expectations[0] = percentile_0;
+        for (int i = 1; i < percentiles.size(); i++) {
+            double percentile_l = percentiles.get(i - 1);
+            double percentile_r = percentiles.get(i);
+            double fractions_l = fractions.get(i - 1);
+            double fractions_r = fractions.get(i);
+            builder.addRange(percentile_l, percentile_r);
+            expectations[i] = (fractions_l * percentile_l + fractions_r * percentile_r) / (fractions_l + fractions_r);
+        }
+        double percentile_n = percentiles.get(percentiles.size() - 1);
+        builder.addUnboundedFrom(percentile_n);
+        expectations[percentiles.size()] = percentile_n;
+        return Tuple.tuple(
+            builder,
+            new BucketCorrelationAggregationBuilder(
+                "correlates",
+                "correlation_range>_count",
+                new CountCorrelationFunction(
+                    new CorrelativeValue(expectations, fractions.stream().mapToDouble(Double::doubleValue).toArray(), totalCount)
+                )
+            )
+        );
+    }
+
+    private double pearsonCorrelation(double[] xs, int[] ys) {
```

Comment:
> I feel a bit uneasy with this test reimplementing the core logic of the calculation to compute the expected value.

That is not what this is doing. This is testing that two mathematical ways of doing the same thing align. One is the true pearson correlation (in the test), the other is our approximation (the aggregation). We could have a single example, but I think that does not offer adequate coverage. 

We may could generate data with regards to a "desired correlation", but it would likely require an equal amount of mathematics. I don't see how that would be any better.

FWIW, this was done with guidance from @tveasey :)

Reply:
This is a tactic I often use in conjunction with fuzzing to verifying implementations. It's particularly useful where there are simple brute force approaches, but even just different just two different implementations are very unlikely to agree on their errors.

Reply:
Ah! OK, cool. Might be worth adding a comment in the test explaining we verify with a different implementation and so on.

Reply:
We do this sort of thing in the aggs tests as well from time to time. We often have a simple example we can assert produces the numbers we expect but we other times fire random numbers into the thing and make an idealized implementation and assert that they are the same. Its nice to make sure none of the distributed "stuff" got in the way. And there really isn't a substitute for randomized data to find weird edges.